module/zfs/zpl_xattr.c

```diff
@@ -1408,6 +1413,655 @@ xattr_handler_t zpl_xattr_acl_default_handler =
 
 #endif /* CONFIG_FS_POSIX_ACL */
 
+#ifdef CONFIG_FS_NFS4_ACL
+
+static const struct cred *nfs4acl_resolver_cred;
+
+#define NFS4ACL_XATTR "system.nfs4_acl"
+#define NFS4ACL_NAMESZ  128
+#define NFS4ACL_UINT_MAXLEN 11
+#define NFS4ACL_XDR_MOD 4
+
+/* XXX The default functions from the Linux kernel for this defined in
+   security/keys/user_defined.c are marked GPL-only; these are
+   logically equivilent but not copied, any issues with that? */
+static int
+nfs4acl_key_preparse(struct key_preparsed_payload *kpp)
+{
+	struct user_key_payload *ukp;
+	size_t dl = kpp->datalen;
+
+	if (!kpp->data || dl < 1 || dl > 32767)
+		return -EINVAL;
+
+	ukp = kmalloc(dl + sizeof(*ukp), GFP_KERNEL);
+	if (!ukp)
+		return -ENOMEM;
+
+	ukp->datalen = dl;
+	kpp->quotalen = dl;
+	kpp->payload.data[0] = ukp;
+	memcpy(ukp->data, kpp->data, dl);
+	return 0;
+}
+
+static void
+nfs4acl_key_free_preparse(struct key_preparsed_payload *kpp)
+{
+	kfree(kpp->payload.data[0]);
+}
+
+static void
+nfs4acl_key_destroy(struct key *k)
+{
+	struct user_key_payload *ukp = k->payload.data[0];
+
+	kfree(ukp);
+}
+
+static void
+nfs4acl_key_describe(const struct key *k, struct seq_file *sf)
+{
+	seq_puts(sf, k->description);
+	if (key_is_positive(k))
+		seq_printf(sf, ": %u", k->datalen);
+}
+
+static long
+nfs4acl_key_read(const struct key *k, char __user *ub, size_t ublen)
+{
+	const struct user_key_payload *ukp;
+	long ret;
+
+	/* XXX function was user_key_payload in older kernels, changed in 4.11
+	       - need to wrap this to do the right thing based on version */
+	ukp = user_key_payload_rcu(k);
+	ret = ukp->datalen;
+
+	if (ub && ublen > 0) {
+		if (ublen > ukp->datalen)
+			ublen = ukp->datalen;
+
+		if (copy_to_user(ub, ukp->data, ublen) != 0)
+			ret = -EFAULT;
+	}
+
+	return ret;
+}
+
+/* Use the existing nfs4 userspace resolver mechanism, but instantiate it
+   under a different name to avoid a conflict with nfs. This will require a new
+   line in /etc/request-key.conf or a new file in /etc/request-key.d containing:
+
+   create  zfs_nfs4acl_resolver    *       *       /usr/sbin/nfsidmap -t 600 %k %d
+
+   It would be better to have a native kernel interface to NFS4 ACLs to avoid
+   dealing with uid/gid mappings in-kernel and the unnecessary local use of the
+   NFS4 domain, but this is also better than not having NFS4 ACL support at all 8-/.
+*/
```

Comment:
So, what did you guys do to make the native ACL not be applied to permission checks? I assume when a pool is imported from another system with native ACLs on it and the linux acltype is noacl or posixacl the system ignores them. While with my current changes I can see and modify the native ACL, it doesn't appear to be used for actual access control:

$ nfs4_getfacl /aclpool/file
A:O:owner@:rwatTnNcCoy
A::henson@pbhware.com:rtncy

$ id
uid=1000(henson) gid=1004(henson)

$ cat /aclpool/file
cat: /aclpool/file: Permission denied

What do I need to do when the acltype is set to nfs4acl to have zfs process them?

Reply:
I sorted out how to plumb the linux VFS into the ZFS access code; just pushed an initial implementation.